### PR TITLE
fix: Lazily elaborate functions in comptime interpreter

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/pull/5538#issuecomment-2237533285

## Summary\*

If a function with an empty body is found within the comptime interpreter we now lazily elaborate it instead of failing with an error.

## Additional Context

The test for this is `noir_test_success_comptime_globals` which after #5538 requires lazy elaboration due to comptime globals themselves now being elaborated before functions.

This PR opens the way to be more lax about marking functions `comptime` and potentially lets us remove the restriction that you can only call non-comptime functions in separate modules.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/elaborator/comptime.rs [hunk 1]
```diff
@@ -11,42 +9,40 @@ impl<'context> Elaborator<'context> {
...
-        let function_id = current_function.map(DependencyId::Function);
-        let old_item = replace(&mut self.current_item, function_id);
+        // Create a fresh elaborator to ensure no state is changed from
+        // this elaborator
+        let mut elaborator = Elaborator::new(
+            self.interner,
+            self.def_maps,
+            self.crate_id,
+            self.debug_comptime_in_file,
+        );
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/comptime.rs:24]: I didn't encounter any errors in reusing the same elaborator but I think this approach is safer and requires less tear-down code.

File: compiler/noirc_frontend/src/elaborator/mod.rs [hunk 2]
```diff
@@ -130,8 +131,6 @@ pub struct Elaborator<'context> {
     /// to the corresponding trait impl ID.
     current_trait_impl: Option<TraitImplId>,
 
-    trait_id: Option<TraitId>,
-
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/mod.rs:134]: `Elaborator::trait_id` was set only while resolving trait impls while `FuncMeta::trait_id` is set only for trait functions. Both are used for resolving trait methods either to a specific impl or to the trait itself - so they were incompatible. I found that we already had `Elaborator::current_trait_impl` for trait impls, and the `TraitId` could be retrieved indirectly through that, so this field was redundant and removing it fixed the conflict.

**Review by michaeljklein — CHANGES_REQUESTED** [review]
Small question r.e. tearing down the fresh elaborator, otherwise LGTM

File: compiler/noirc_frontend/src/elaborator/comptime.rs [hunk 3]
```diff
@@ -11,42 +9,40 @@ impl<'context> Elaborator<'context> {
...
+        let result = f(&mut elaborator);
+        elaborator.check_and_pop_function_context();
 
-        self.generics = old_generics;
-        self.current_item = old_item;
-        self.scopes.end_function();
-        self.scopes.0.push(old_scope);
-        self.check_and_pop_function_context();
+        self.comptime_scopes = elaborator.comptime_scopes;
+        self.errors.append(&mut elaborator.errors);
```
[the remark this answers is not in the document]
Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/comptime.rs:45]: We shouldn't have to check those two:
- `resolving_ids`: Only set & reset while resolving a struct. So even if we do resolve a struct item here, the added id would be removed when it is finished.
- `unresolved_globals`: Only set `elaborate_items` and similarly later cleared within the same function.

**Review by michaeljklein — APPROVED** [review]

**Pull request merged**